eval.go

```diff
@@ -450,34 +450,47 @@ func (d *indexData) List(ctx context.Context, q query.Q) (rl *RepoList, err erro
 		tr.Finish()
 	}()
 
+	var include func(rle *RepoListEntry) (bool, error)
+
 	q = d.simplify(q)
 	tr.LazyLog(q, true)
 	if c, ok := q.(*query.Const); ok {
-		return d.maybeRepoList(c.Value), nil
-	}
-
-	sr, err := d.Search(ctx, q, &SearchOptions{
-		ShardMaxMatchCount: 1,
-		TotalMaxMatchCount: 1,
-	})
-	if err != nil {
-		return nil, err
+		if !c.Value {
+			return &RepoList{}, nil
+		}
+		include = func(rle *RepoListEntry) (bool, error) {
+			return true, nil
+		}
+	} else {
+		// We need to run a search per repo to decide if it is included.
+		include = func(rle *RepoListEntry) (bool, error) {
+			qOneRepo := query.NewAnd(
+				query.NewRepoSet(rle.Repository.Name),
+				q)
+			sr, err := d.Search(ctx, qOneRepo, &SearchOptions{
+				ShardMaxMatchCount: 1,
+				TotalMaxMatchCount: 1,
+			})
+			if err != nil {
+				return false, err
+			}
+			return len(sr.Files) > 0, nil
+		}
 	}
 
-	return d.maybeRepoList(len(sr.Files) > 0), nil
-}
-
-// maybeRepoList returns a singleton repo list containing the repo for d if
-// include is true. Otherwise it returns an empty list.
-func (d *indexData) maybeRepoList(include bool) *RepoList {
-	l := &RepoList{}
-	if include {
-		l.Repos = make([]*RepoListEntry, 0, len(d.repoListEntry))
-		for _, rle := range d.repoListEntry {
-			l.Repos = append(l.Repos, &rle)
+	repos := make([]*RepoListEntry, 0, len(d.repoListEntry))
+	for _, rle := range d.repoListEntry {
+		ok, err := include(&rle)
+		if err != nil {
+			return nil, err
+		}
+		if ok {
+			rleCopy := rle
```

Comment:
Ahah, I was fixing this exact bug as part of the List changes I was doing. How about we return a pointer to the original memory location in repoListEntry rather than a copy? It should be short lived and read only right?

Reply:
I added a test in my PR.

Reply:
Thanks. Just in case I'll keep the copy. Things that "escape" the eval layer are copied. This is mostly done due to references to read-only MMAP regions, but seems like a good thing to do in general so will keep doing it for this part of the code. Even though I think it is safe to not copy? Either way this copy isn't a true copy, since things like the subrepo maps won't be copied.

I'll leave it as is, but happy to change it in a follow up.